### PR TITLE
fix: Add polyfill for Chrome V65+

### LINKED ISSUE #23509: cypress 10.4+ in chrome 70 breaks with globalThis
### Current behavior

Cypress run on chrome 70 cannot start with error: "Uncaught ReferenceError: globalThis is not defined"

### Desired behavior

cypress run with chrome 70

### Test code to reproduce

Any test with cypress 10.4+ run on chrome 70 should reproduce.

### Cypress Version

10.4+

### Node version

v16.5.0

### Operating System

linux

### Debug Logs

```shell
root@me-cd-test-k8s-3137-446pp-dqxkt-wgg0b:/home/jenkins/agent/workspace/ME-CD/test-k8s# DEBUG=cypress:launcher:* node test.js cypress/prepare/role.cy.js /tmp/test-result/3137/admin
✔  Verified Cypress! /root/.cache/Cypress/10.4.0/Cypress
[16160:0823/155424.906310:ERROR:zygote_host_impl_linux.cc(263)] Failed to adjust OOM score of renderer with pid 16344: Permission denied (13)
[16344:0823/155425.028386:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.
2022-08-23T07:54:28.962Z cypress:launcher:detect detecting if the following browsers are present [ { name: 'chrome', family: 'chromium', channel: 'stable', displayName: 'Chrome', versionRegex: /Google Chrome (\S+)/m, binary: [ 'google-chrome', 'chrome', 'google-chrome-stable' ], minSupportedVersion: 64 }, { name: 'chromium', family: 'chromium', channel: 'stable', displayName: 'Chromium', versionRegex: /Chromium (\S+)/m, binary: [ 'chromium-browser', 'chromium' ], minSupportedVersion: 64 }, { name: 'chrome', family: 'chromium', channel: 'beta', displayName: 'Chrome Beta', versionRegex: /Google Chrome (\S+) beta/m, binary: 'google-chrome-beta', minSupportedVersion: 64 }, { name: 'chrome', family: 'chromium', channel: 'canary', displayName: 'Canary', versionRegex: /Google Chrome Canary (\S+)/m, binary: 'google-chrome-canary', minSupportedVersion: 64 }, { name: 'firefox', family: 'firefox', channel: 'stable', displayName: 'Firefox', versionRegex: /^Mozilla Firefox ([^\sab]+)$/m, binary: 'firefox', minSupportedVersion: 86 }, { name: 'firefox', family: 'firefox', channel: 'dev', displayName: 'Firefox Developer Edition', versionRegex: /^Mozilla Firefox (\S+b\S*)$/m, binary: [ 'firefox-developer-edition', 'firefox' ], minSupportedVersion: 86 }, { name: 'firefox', family: 'firefox', channel: 'nightly', displayName: 'Firefox Nightly', versionRegex: /^Mozilla Firefox (\S+a\S*)$/m, binary: [ 'firefox-nightly', 'firefox-trunk' ], minSupportedVersion: 86 }, { name: 'edge', family: 'chromium', channel: 'stable', displayName: 'Edge', versionRegex: /Microsoft Edge (\S+)/m, binary: [ 'edge', 'microsoft-edge' ], minSupportedVersion: 79 }, { name: 'edge', family: 'chromium', channel: 'canary', displayName: 'Edge Canary', versionRegex: /Microsoft Edge Canary (\S+)/m, binary: 'edge-canary', minSupportedVersion: 79 }, { name: 'edge', family: 'chromium', channel: 'beta', displayName: 'Edge Beta', versionRegex: /Microsoft Edge Beta (\S+)/m, binary: 'edge-beta', minSupportedVersion: 79 }, { name: 'edge', family: 'chromium', channel: 'dev', displayName: 'Edge Dev', versionRegex: /Microsoft Edge Dev (\S+)/m, binary: [ 'edge-dev', 'microsoft-edge-dev' ], minSupportedVersion: 79 } ]
Missing baseUrl in compilerOptions. tsconfig-paths will be skipped
2022-08-23T07:54:43.180Z cypress:launcher:detect detecting if the following browsers are present [ { name: 'chrome', family: 'chromium', channel: 'stable', displayName: 'Chrome', versionRegex: /Google Chrome (\S+)/m, binary: [ 'google-chrome', 'chrome', 'google-chrome-stable' ], minSupportedVersion: 64 }, { name: 'chromium', family: 'chromium', channel: 'stable', displayName: 'Chromium', versionRegex: /Chromium (\S+)/m, binary: [ 'chromium-browser', 'chromium' ], minSupportedVersion: 64 }, { name: 'chrome', family: 'chromium', channel: 'beta', displayName: 'Chrome Beta', versionRegex: /Google Chrome (\S+) beta/m, binary: 'google-chrome-beta', minSupportedVersion: 64 }, { name: 'chrome', family: 'chromium', channel: 'canary', displayName: 'Canary', versionRegex: /Google Chrome Canary (\S+)/m, binary: 'google-chrome-canary', minSupportedVersion: 64 }, { name: 'firefox', family: 'firefox', channel: 'stable', displayName: 'Firefox', versionRegex: /^Mozilla Firefox ([^\sab]+)$/m, binary: 'firefox', minSupportedVersion: 86 }, { name: 'firefox', family: 'firefox', channel: 'dev', displayName: 'Firefox Developer Edition', versionRegex: /^Mozilla Firefox (\S+b\S*)$/m, binary: [ 'firefox-developer-edition', 'firefox' ], minSupportedVersion: 86 }, { name: 'firefox', family: 'firefox', channel: 'nightly', displayName: 'Firefox Nightly', versionRegex: /^Mozilla Firefox (\S+a\S*)$/m, binary: [ 'firefox-nightly', 'firefox-trunk' ], minSupportedVersion: 86 }, { name: 'edge', family: 'chromium', channel: 'stable', displayName: 'Edge', versionRegex: /Microsoft Edge (\S+)/m, binary: [ 'edge', 'microsoft-edge' ], minSupportedVersion: 79 }, { name: 'edge', family: 'chromium', channel: 'canary', displayName: 'Edge Canary', versionRegex: /Microsoft Edge Canary (\S+)/m, binary: 'edge-canary', minSupportedVersion: 79 }, { name: 'edge', family: 'chromium', channel: 'beta', displayName: 'Edge Beta', versionRegex: /Microsoft Edge Beta (\S+)/m, binary: 'edge-beta', minSupportedVersion: 79 }, { name: 'edge', family: 'chromium', channel: 'dev', displayName: 'Edge Dev', versionRegex: /Microsoft Edge Dev (\S+)/m, binary: [ 'edge-dev', 'microsoft-edge-dev' ], minSupportedVersion: 79 } ]

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        10.4.0                                                                         │
  │ Browser:        Chrome 70 (headless)                                                           │
  │ Node Version:   v16.5.0 (/usr/local/bin/node)                                                  │
  │ Specs:          1 found (role.cy.js)                                                           │
  │ Searched:       cypress/prepare/role.cy.js                                                     │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  role.cy.js                                                                      (1 of 1)
[cypress-log-to-output] Attempting to connect to Chrome Debugging Protocol
2022-08-23T07:54:44.983Z cypress:launcher:browsers launching browser { browser: { name: 'chrome', family: 'chromium', channel: 'stable', displayName: 'Chrome', version: '70.0.3538.77', path: 'google-chrome', minSupportedVersion: 64, majorVersion: 70, isHeadless: true, isHeaded: false }, url: 'about:blank' }
2022-08-23T07:54:44.985Z cypress:launcher:browsers spawning browser with args { args: [ 'about:blank', '--test-type', '--ignore-certificate-errors', '--start-maximized', '--silent-debugger-extension-api', '--no-default-browser-check', '--no-first-run', '--noerrdialogs', '--enable-fixed-layout', '--disable-popup-blocking', '--disable-password-generation', '--disable-single-click-autofill', '--disable-prompt-on-repos', '--disable-background-timer-throttling', '--disable-renderer-backgrounding', '--disable-renderer-throttling', '--disable-backgrounding-occluded-windows', '--disable-restore-session-state', '--disable-new-profile-management', '--disable-new-avatar-menu', '--allow-insecure-localhost', '--reduce-security-for-testing', '--enable-automation', '--disable-print-preview', '--disable-device-discovery-notifications', '--autoplay-policy=no-user-gesture-required', '--disable-site-isolation-trials', '--metrics-recording-only', '--disable-prompt-on-repost', '--disable-hang-monitor', '--disable-sync', '--disable-web-resources', '--safebrowsing-disable-download-protection', '--disable-client-side-phishing-detection', '--disable-component-update', "--simulate-outdated-no-au='Tue, 31 Dec 2099 23:59:59 GMT'", '--disable-default-apps', '--use-fake-ui-for-media-stream', '--use-fake-device-for-media-stream', '--disable-ipc-flooding-protection', '--disable-backgrounding-occluded-window', '--disable-breakpad', '--password-store=basic', '--use-mock-keychain', '--disable-dev-shm-usage', '--disable-gpu', '--no-sandbox', '--proxy-server=http://localhost:39355', '--headless', '--window-size=1280,720', '--force-device-scale-factor=1', '--remote-debugging-port=33382', '--remote-debugging-address=127.0.0.1', '--user-data-dir=/root/.config/Cypress/cy/production/browsers/chrome-stable/run-16160', '--disk-cache-dir=/root/.config/Cypress/cy/production/browsers/chrome-stable/run-16160/CypressCache' ] }
2022-08-23T07:54:45.198Z cypress:launcher:browsers chrome stderr: [0823/155445.197221:ERROR:gpu_process_transport_factory.cc(980)] Lost UI shared context.
2022-08-23T07:54:45.200Z cypress:launcher:browsers chrome stderr: DevTools listening on ws://127.0.0.1:33382/devtools/browser/13cdca87-00d4-4c07-903c-5064f790ef5a
2022-08-23T07:54:45.208Z cypress:launcher:browsers chrome stderr: [0823/155445.206777:ERROR:zygote_host_impl_linux.cc(259)] Failed to adjust OOM score of renderer with pid 16474: Permission denied (13)
[cypress-log-to-output] Connected to Chrome Debugging Protocol
2022-08-23T07:54:46.896Z cypress:launcher:browsers chrome stderr: [0823/155446.895003:INFO:CONSOLE(39990)] "Uncaught ReferenceError: globalThis is not defined", source: http://me-e2e-cypress10.me.magic.com/__/assets/index.23f3e8ba.js (39990)
```

### Other

globalThis is not supported in chrome 70, 10.4+ somehow use it directly in bundled code.

### PR DESCRIPTION
**Pull request by rockindahizzy**
- Closes #23509

### User facing changelog

- Fix regression causing incompatibility with Chrome 70 and older

### Steps to test

Try to run e2e tests (I used the kitchen sink app) using Chrome 70.
Prior to this change, nothing will render in the UI and tests will timeout.
After the fix, tests will run normally.
I can provide the chrome binary (macOS) if needed.

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- na Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- na Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by rockindahizzy** [comment]
I'm not sure if there is a way/need to test this fix

**Review by lmiller1990 — APPROVED** [review]
Nice, looks good to me. We should probably have Kitchensink running on our lowest supported Chrome for every branch.

For anyone interested, this is how the plugin works:

```html
<script type="module">try{import.meta.url;import("_").catch(()=>1);}catch(e){}window.__vite_is_modern_browser=true;</script>
  <script type="module">!function(){if(window.__vite_is_modern_browser)return;console.warn("vite: loading legacy build because dynamic import or import.meta.url is unsupported, syntax error above should be ignored");var e=document.getElementById("vite-legacy-polyfill"),n=document.createElement("script");n.src=e.src,n.onload=function(){System.import(document.getElementById('vite-legacy-entry').getAttribute('data-src'))},document.body.appendChild(n)}();</script>
```

It makes two entries, and only serves the legacy if the first one fails. "Legacy" is `try{import.meta.url;import("_").catch(()=>1);}` -  either `import.meta.url` doesn't exist or `import()` isn't defined.

This means we ship two bundlers - legacy and non legacy - so the binary is a couple hundred kb extra, not ideal, but probably the only way to support older browsers :shrug:.

**Review by mike-plummer — CHANGES_REQUESTED** [review]
I think a bit more work is needed here. there's more than just the `globalThis` issue, that's just the first one that got hit. I get the following error and the SpecsList fails to render (sidebar does though) in Chromium v70 - `replaceAll` is Chrome 85+

Screen Shot 2022-09-07 at 9 00 08 AM

**Comment by lmiller1990** [comment]
Right @mike-plummer good point - I was just reviewing for the globalThis polyfill.

Since we specify `esnext` here, TS won't warn us. I don't think we should use `esnext` (at all) since it assumes our target environment supports the latest JS features, like `replaceAll`, which it doesn't.

Line: https://github.com/cypress-io/cypress/blob/develop/packages/frontend-shared/tsconfig.json#L6 

The confusing thing is our minimal Node.js version has different features to our minimum Chrome version. Ideally, we'd have two `tsconfig.json` we extend from: one for frontend code and one for backend code. Those would be configured in a way that ensures we cannot write code that isn't supported.

For now, I think we should change the main `tsconfig.json` for frontend code (`frontend-shared/tsconfig.json`) to not target `esnext`, but something lower (`es6` aka `es2015` is probably fine). Then TS should give us compilation errors, for example telling us that `replaceAll` isn't okay. 

Edit: I just tried this in a simple project and it works, but **not** in the monorepo - probably since there's a bunch of `tsconfig.json` files and it's not clear how/which one is getting consumed during compilation? So some manual work or tinkering is probably required.

**Review by lmiller1990 — COMMENTED** [review]
Left some thoughts about how we might modify/change tsconfig to ensure we don't write invalid code in the future - might require some additional work.

**Comment by rockindahizzy** [comment]
@lmiller1990, if we set it to es2015, will we even need to polyfill?

**Comment by lmiller1990** [comment]
> @lmiller1990, if we set it to es2015, will we even need to polyfill?

Yes we do - here's why. `lib: es2015` only impacts language *features* (think syntax) - not additional/new methods. Eg:

```json
{
  "compilerOptions": {
    "target": "ES5",
    "lib": ["ES2021"]
  },
  "include": ["./index.ts"]
}
```

```ts
// index.ts
"AAA".replaceAll("A", "B")

export function blah () {
  return () => 'AAA'
}
```

`yarn tsc -p .` gives:

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blah = void 0;
"AAA".replaceAll("A", "B");
function blah() {
    return function () { return 'AAA'; };
}
exports.blah = blah;
```

Notice it will transform the `() => 'AAA'` to `function` (no arrow functions in `ES5`) but `replaceAll` is compiled as is. 

One argument for this behavior is because it's possible for a user to do `String.prototype.replaceAll = ...`, and TS would have no way to know. So, the compiler only compiles features.

If we want to polyfill old methods, we'd need to use babel or to include a polyfill.

**Review by lmiller1990 — COMMENTED** [review]
Commented!

**Comment by tgriesser** [comment]
> Nice, looks good to me. We should probably have Kitchensink running on our lowest supported Chrome for every branch.

@lmiller1990 is this defined anywhere? I could be wrong but I believe Chrome 65 was EOL in 2018:

https://endoflife.software/applications/browsers/google-chrome

**Comment by lmiller1990** [comment]
> > Nice, looks good to me. We should probably have Kitchensink running on our lowest supported Chrome for every branch.
> 
> @lmiller1990 is this defined anywhere? I could be wrong but I believe Chrome 70 was EOL in 2018:
> 
> https://endoflife.software/applications/browsers/google-chrome

Hm, the issue says we should add the polyfill https://github.com/cypress-io/cypress/issues/23509. Which leads me to assume we support Chrome <70?

I think we probably need to actually define what minimum browser versions we support, and make sure that is running on Kitchensink (eg - if not Chrome 70, than what version). See [thread](https://cypressio.slack.com/archives/C011BAPC614/p1662666994632629) - I'll also tag you.

**Comment by rockindahizzy** [comment]
We support the following browser versions [according to the docs](https://docs.cypress.io/guides/guides/launching-browsers#Browser-versions-supported)

- Chrome 64 and above.
- Edge 79 and above.
- Firefox 86 and above.

**Comment by rockindahizzy** [comment]
> I think a bit more work is needed here. there's more than just the `globalThis` issue, that's just the first one that got hit. I get the following error and the SpecsList fails to render (sidebar does though) in Chromium v70 - `replaceAll` is Chrome 85+
> 
> Screen Shot 2022-09-07 at 9 00 08 AM

@mike-plummer, I am now polyfilling all modules. `replaceAll` should be working now

File: packages/app/vite.config.mjs [hunk 1]
```diff
@@ -29,5 +30,9 @@ export default makeConfig({
         dest: 'dist',
       }],
     }),
+    Legacy({
+      targets: ['Chrome >= 64', 'Firefox >= 86', 'Edge >= 79'],
+      modernPolyfills: true,
```
Comment (mike-plummer) [inline packages/app/vite.config.mjs:35]: Why enable this option? The docs seem to recommend against it. I think all we care about is the `polyfills` option which is `true` by default and auto-generated off the `targets` entries

Comment (rockindahizzy) [reply packages/app/vite.config.mjs:35]: It didn't work for me without this option set 😕

Comment (mike-plummer) [reply packages/app/vite.config.mjs:35]: Huh, interesting. I must be misunderstanding the intent of that field.

**Review by mike-plummer — APPROVED** [review]
LGTM, tested on Chromium v70

**Review by astone123 — APPROVED** [review]
Worked for me 👍🏻

**Review by ZachJW34 — APPROVED** [review]
Paired with @rockindahizzy, works great!

**Comment by rockindahizzy** [comment]
@lmiller1990, I looked into the tsconfig changes a little more and am thinking we break that change into another issue to avoid blowing up this issue and not getting the fix into this release.

Also, I'm still having a hard time understanding what changing the tsconfig really adds since we are relying on vite to polyfill for us.

**Comment by lmiller1990** [comment]
@rockindahizzy I don't think we need to change tsconfig anymore - if we are shipping the entire polyfill, it should be fine. I was more thinking "can we avoid shipping a huge polyfill" but I don't think this is an option until we officially drop support for older browsers.

**Pull request merged**